web/autolinker.js

```diff
@@ -0,0 +1,149 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  AnnotationBorderStyleType,
+  AnnotationType,
+  createValidAbsoluteUrl,
+  Util,
+} from "../src/shared/util.js";
+import { getOriginalIndex, normalize } from "./pdf_find_controller.js";
+
+function DOMRectToPDF({ width, height, left, top }, pdfPageView) {
+  if (width === 0 || height === 0) {
+    return null;
+  }
+
+  const pageBox = pdfPageView.textLayer.div.getBoundingClientRect();
+  const bottomLeft = pdfPageView.getPagePoint(
+    left - pageBox.left,
+    top - pageBox.top
+  );
+  const topRight = pdfPageView.getPagePoint(
+    left - pageBox.left + width,
+    top - pageBox.top + height
+  );
+
+  return Util.normalizeRect([
+    bottomLeft[0],
+    bottomLeft[1],
+    topRight[0],
+    topRight[1],
+  ]);
+}
+
+function calculateLinkPosition(range, pdfPageView) {
+  const rangeRects = range.getClientRects();
+  if (rangeRects.length === 1) {
+    return { rect: DOMRectToPDF(rangeRects[0], pdfPageView) };
+  }
+
+  const rect = [Infinity, Infinity, -Infinity, -Infinity];
+  const quadPoints = [];
+  let i = 0;
+  for (const domRect of rangeRects) {
+    const normalized = DOMRectToPDF(domRect, pdfPageView);
+    if (normalized === null) {
+      continue;
+    }
+
+    quadPoints[i] = quadPoints[i + 4] = normalized[0];
+    quadPoints[i + 1] = quadPoints[i + 3] = normalized[3];
+    quadPoints[i + 2] = quadPoints[i + 6] = normalized[2];
+    quadPoints[i + 5] = quadPoints[i + 7] = normalized[1];
+
+    rect[0] = Math.min(rect[0], normalized[0]);
+    rect[1] = Math.min(rect[1], normalized[1]);
+    rect[2] = Math.max(rect[2], normalized[2]);
+    rect[3] = Math.max(rect[3], normalized[3]);
+
+    i += 8;
+  }
+  return { quadPoints, rect };
+}
+
+function createLinkAnnotation({ url, index, length }, pdfPageView) {
+  const convertedMatch = pdfPageView._textHighlighter._convertMatches(
+    [index],
+    [length]
+  )[0];
+
+  const range = new Range();
+  range.setStart(
+    pdfPageView._textHighlighter.textDivs[convertedMatch.begin.divIdx]
+      .firstChild,
+    convertedMatch.begin.offset
+  );
+  range.setEnd(
+    pdfPageView._textHighlighter.textDivs[convertedMatch.end.divIdx].firstChild,
+    convertedMatch.end.offset
+  );
+
+  return {
+    unsafeUrl: url,
+    url,
+    annotationType: AnnotationType.LINK,
+    rotation: 0,
+    ...calculateLinkPosition(range, pdfPageView),
+    // This is just the default for AnnotationBorderStyle. At some point we
+    // should switch to something better like `new LinkAnnotation` here.
```

Comment:
Given that `new LinkAnnotation` is *worker-thread functionality* there's simply no good/efficient way to use that here, so let's drop that part of the comment.
```suggestion
    // This is just the default for AnnotationBorderStyle.
```